android/cli/hooks/run.js

```diff
@@ -233,44 +233,56 @@ exports.init = function (logger, config, cli) {
 
 				function (next) {
 					var logBuffer = [],
+						lastLogLevel,
 						displayStartLog = true,
 						tiapiRegExp = /^(\w\/TiAPI\s*\:)/,
 						instances = deviceInfo.length,
 						endLog = false;
 
 					function printData(device, deviceName, line) {
+						var logLevel = lastLogLevel; // if continuing from middle of last message, keep same log level
+
+						// start of a new log message
 						if (device.appPidRegExp.test(line)) {
 							line = line.trim().replace(/\%/g, '%%').replace(device.appPidRegExp, ':');
-							var logLevel = line.charAt(0).toLowerCase();
+							logLevel = line.charAt(0).toLowerCase();
 							if (tiapiRegExp.test(line)) {
 								line = line.replace(tiapiRegExp, '').trim();
 							} else {
 								line = line.replace(/^\w\/(\w+)\s*\:/g, '$1:').grey;
 							}
 							line = deviceName + line;
-							switch (logLevel) {
-								case 'v':
-									logger.trace(line);
-									break;
-								case 'd':
-									logger.debug(line);
-									break;
-								case 'w':
-									logger.warn(line);
-									break;
-								case 'e':
-									logger.error(line);
-									break;
-								case 'i':
-								default:
-									logger.info(line);
+						} else {
+							// if it begins with something like "E/SQLiteLog( 1659):" it's not a contination, don't log it.
+							if (/^\w\/\w+\s*\(\s*\d+\):/.test(line)) {
```

Comment:
Yes and no. Main solution is in adb.js, using stream splitter to pump out events. That library basically enforces we spit an event per-line and not mid-line. This is an additional catch that if a log message gets spread across multiple lines we'll still log it (unless it looks like it's the start of another log message that isn't TiAPI).